### DISCUSSION TITLE
Fortran test function is not consistent with spacepy.irbempy.get_Bfield()

### DISCUSSION THREAD

**Discussion by anthony2352** [spacepy/spacepy#579] [Support and Q&A]
Magnetic field value from Fortran 77 test function that calls T04_s subroutine directly is not consistent with spacepy.irbempy.get_Bfield().

### Situation:

I have written the following python code:
```
import datetime as dtm
import spacepy as sp
import spacepy.coordinates as spc
import spacepy.time as spt
import spacepy.irbempy as ir
import numpy as np

'''
omni = {'Dst':   ,
        'Pdyn':  ,
        'ByIMF': ,
        'BzIMF': ,
        'W'    : [  ]
       }
'''
vals = [0.0, -10.0, 0.0, 0.0, 2.97,-16.600000, -8.200000, 0.0, 0.0, 0.0, 1.022, 0.874,  0.807,  1.657000,  2.737,  4.542000]
keys = ['Kp', 'Dst', 'dens', 'velo', 'Pdyn', 'ByIMF', 'BzIMF', 'G1', 'G2', 'G3', 'W1', 'W2', 'W3', 'W4', 'W5', 'W6']
dic = {k:[v] for k,v in zip(keys, vals)}
t = spt.Ticktock(['2013-04-01T12:59:30'],'ISO')
y = spc.Coords([4, 1.0e-05, -1.0e-05],'GSM', 'car')
b_field = ir.get_Bfield(t,y,extMag='T05',omnivals=dic)
print(b_field)

```
Here, I am creating a dictionary containing some omni values. As I understand, the parameters Kp, dens, velo, G1, G2, and G3 are not used in the T05 model and do not impact the result, so I left them as 0. I previously used spacepy.omni.get_omni() to look up these values for the date and time 2003-10-29T06:00:00. However, for the external magnetic field model, it should not matter what the date is, correct? Or, is this required to calculate the geodipole tilt angle? If not, the date and time only seems necessary to calculate the IGRF, which is what Geopack does. As you can see from the code, I am passing the location (4.0, 1.0e-05, and -1.0e-05) in GSM and in cartesian coordinate system. The problem that I am having is that I wrote some Fortran code that calls the subroutine t04_s in T05 directly and passed the exact same parameters, but I got a completely different magnetic field vector. I am attaching the my Fortran test function and T05.for. For the geodipole tilt angle, I set it as 0.170481. I wasn't sure how to obtain this using spacepy. If the geodipole tilt angle is passed to spacepy.irbempy.get_Bfield() based on the date, then this very well may be the reason for the disagreement. I believe that I am either passing that parameters incorrectly into spacepy, or I am interpreting the output incorrectly. Is there any way that I can resolve this disagreement? Here are the fortran codes (It would not actually let me add a fortran document, so they are both text files. You will need to convert them to fortran files):
[get_bfield_main.txt](https://github.com/spacepy/spacepy/files/6896974/get_bfield_main.txt)
[T05.txt](https://github.com/spacepy/spacepy/files/6896975/T05.txt)
 *Edit: I changed the coordinate system from 'GSE' to 'GSM'. I realized that T05 expects GSM; however, this did not resolve the disagreement.

### OS, Python version, and dependency version information:

sys.version_info(major=3, minor=8, micro=5, releaselevel='final', serial=0)
numpy=1.19.2
scipy=1.5.2
matplotlib=3.3.2

### Version of SpacePy

I cloned spacepy from the git repository, and it is version 0.2.2.

**Comment by drsteve** [comment]
Hi, apologies for the late response - I guess I missed seeing this when it was posted. Anyway, while you may well have figured things out by now it's probably good to have a record of the resolution here in case someone else has a similar question. (Hence I've also move this to discussion, as it's neither a SpacePy issue nor an IRBEM issue.)

There are a few reasons why the answer would disagree in this case:
1. As noted in the header, the `t04_s` routine provides the _external_ field. That is, it explicitly doesn't include the Earth's main field. So the IGRF field (or whatever internal model you want) at the point being evaluated needs to be added back on. This, depending on where you are evaluating the model, can be a very significant effect on both magnitude and direction.
2. `get_Bfield` wraps the `GET_FIELD1` subroutine in the IRBEM library. These do a number of things to get the total vector B (internal + external field), including necessary coordinate conversions (so it doesn't matter what input system you use as long as it's correctly specified), initialization of transforms and IGRF (which includes calculating the dipole axis), evaluation of the internal and external fields, ...
3. The multi-purpose magnetic field evaluation routine in IRBEM is `CHAMP` (French for "field"), and this is called after all the setup is done in `GET_FIELD1`. Notably, the magnetic field vectors are converted back to Cartesian geographic coordinates in all cases before being returned, which will also affect the field vector (although this won't affect the magnitude).

None of our release versions have exposed the functionality to return the dipole tilt angle, although this should be available in the future. That will, as noted in the question, make a difference to the returned value, but the real difference is that the things being compared are not the same and are not expected to be the same.
